src/c2goto/library/semaphore_lib.c

```diff
@@ -0,0 +1,106 @@
+#include <semaphore.h>
+
+#define __ESBMC_sem_lock_field(a) ((a).__lock)
+#define __ESBMC_sem_count_field(a) ((a).__count)
+#define __ESBMC_sem_init_field(a) ((a).__init)
+
+/************************* Sem manipulation routines ************************/
+
+int sem_init(sem_t *__sem, int __pshared, unsigned int __value)
+{
+  __ESBMC_atomic_begin();
+  __ESBMC_assert(!__pshared, "ESBMC does not check multiprocesses for now");
+  __ESBMC_sem_lock_field(*__sem) = 0;
+  __ESBMC_sem_count_field(*__sem) = __value;
+  __ESBMC_sem_init_field(*__sem) = 1;
+  __ESBMC_atomic_end();
+  return 0;
+}
+
+int sem_init_check(sem_t *__sem)
+{
+  // check whether this sem has been initialised
+  __ESBMC_atomic_begin();
+  __ESBMC_assert(__ESBMC_sem_init_field(*__sem), "Sem is not initialised");
+  __ESBMC_atomic_end();
+  return 0;
+}
+
+int sem_wait_check(sem_t *__sem)
+{
+  __ESBMC_atomic_begin();
+  sem_init_check(__sem);
+  __ESBMC_sem_count_field(*__sem) -= 1;
+  __ESBMC_assert(
+    __ESBMC_sem_count_field(*__sem) < 0, "Deadlocked state in sem_wait");
```

Comment:
Maybe I'm misunderstanding, but this assertion seems strange to me. Initializing a semaphore with value 2 and calling sem_wait() once would trigger this assertion, right?

Reply:
Hi @fbrausse, the purpose of this assertion is to check for deadlock caused by multiple calls to sem_wait(). According to your example, if value is 2, it will trigger this assertion when we call `sem_wait()` for the **third** time in a row. Because our semaphore is decremented each time the function is called.

It is designed to avoid such situations:
```
void *increment(void *arg)
{
  sem_wait(&sem); // decrement the semaphore
  count++;
  sem_wait(&sem); // decrement the semaphore again and it will trigger this assertion, may cause deadlock
  return NULL;
}
...

sem_init(&sem, 0, 1);
```